### PR TITLE
RE-1317 Add basic lint script

### PR DESCRIPTION
**Pull request by mattt416**
This commit adds a basic PR job that lint checks changes being submit
to the repo. To summarise, it:

1. Checks the number of components being changed at the same time
   (more than 1 is a problem).
2. Ensures that a component change is not happening along-side a non
   component change (components should be registered / released in
   isolation).
3. If a single component is being updated, it:
   a) Checks if it's a new component and validates the schema.
   b) Checks if it's a new version release and validates the schema.

### CODE REVIEW

File: gating/pre_merge_check/run [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+set +x; . .venv3/bin/activate; set -x
+pip install -c constraints.txt -r requirements.txt
+pushd rpc_component
+  pip install .
+popd
+
+component_count=0
+non_component_count=0
+
+changes=$(git log --pretty=format: --name-only --no-merges -1)
```
Comment (git-harry) [inline gating/pre_merge_check/run]: Is this assuming a pull request only has one commit?

File: gating/pre_merge_check/run [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+for change in ${changes}; do
+  if [[ "${change}" =~ "components/" ]]; then
```
Comment (git-harry) [inline gating/pre_merge_check/run]: I'd add a `^` to ensure it matches from the start of the string, `if [[ "${change}" =~ ^"components/" ]]; then`

File: rpc_component/rpc_component/schemata.py [hunk 3]
```diff
@@ -166,6 +166,18 @@ def is_series_names_unique(releases):
     }
 )
 
+comparison_added_component_schema = Schema(
+    And(
+        {
+            And(str, len): {
+                "added": And(component_schema),
```
Comment (git-harry) [inline rpc_component/rpc_component/schemata.py:173]: `component_schema` will match against components with releases defined. If were going to allow the population of a new component with existing releases then that's fine (it is likely we'll want to import old releases) but we'll probably want to update `component` to support that (I'm not advocating that be done in this patch).

File: gating/pre_merge_check/run (hunk 1 above)
Comment (mattt416) [reply gating/pre_merge_check/run]: Good point -- switched this to using git diff so we can diff between `master` and `HEAD`.  Any reason that wouldn't work?

File: gating/pre_merge_check/run (hunk 2 above)
Comment (mattt416) [reply gating/pre_merge_check/run]: Oops, just realised your carot is outside the quote ... couldn't get it to work inside quotes so dropped the quotes all together. :)

Comment (git-harry) [reply gating/pre_merge_check/run]: If it's inside the quotes it's treated as a string, I should made that clearer.

File: gating/pre_merge_check/run (hunk 1 above)
Comment (git-harry) [reply gating/pre_merge_check/run]: Sounds reasonable. My only thought would be to use variables so they only get defined in one place and use the names to make it clear what they represent. Reading the script, it's not obvious that it is expecting HEAD to be the merge branch for the pull request.

File: gating/pre_merge_check/run [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  && { echo '[FAIL] A component and non-component change was submitted at the same time'; RC=1; } \
+  || echo '  [OK] A component and non-component change was not submitted at the same time'
+
+if [[ ${component_count} -eq 1 ]]; then
+  # The component cli tool takes a `from` and `to` SHA. Since we don't know how
+  # many commits are in the PR, we use `master` as the `from` SHA and `HEAD` as
+  # the `to` SHA. This allows us to compare against the last committed commit
+  # and the changes proposed in the PR.
+  src_branch="master"
+  pr_branch="HEAD"
```
Comment (git-harry) [inline gating/pre_merge_check/run]: I'd say move this to before line 15 so it can make use of this as well.

Comment (mattt416) [reply gating/pre_merge_check/run]: 👍

**Review by git-harry — APPROVED** [review]

**Pull request merged**